src/module/actor/sheet/popups/casting-item-create-dialog.ts

```diff
@@ -0,0 +1,105 @@
+import { ActorPF2e } from "@actor/index";
+import { SpellPF2e } from "@item";
+import { SpellConsumableItemType } from "@item/consumable/spell-consumables";
+import { OneToTen } from "@module/data";
+import { ErrorPF2e } from "@util";
+
+type FormInputData = {
+    itemTypeOptions?: Object;
+    validLevels?: number[];
+    itemType?: SpellConsumableItemType;
+    level?: OneToTen;
+};
+
+type FormOutputData = {
+    itemType: SpellConsumableItemType;
+    level: OneToTen;
+};
+
+const itemTypeOptions = Object.fromEntries(
+    new Map<SpellConsumableItemType, string>([
+        ["scroll", "PF2E.CastingItemCreateDialog.scroll"],
+        ["wand", "PF2E.CastingItemCreateDialog.wand"],
+        ["cantripDeck5", "PF2E.CastingItemCreateDialog.cantripDeck5"],
+    ])
+);
+
+export class CastingItemCreateDialog extends FormApplication<ActorPF2e> {
+    onSubmitCallback: CastingItemCreateCallback;
+    spell: SpellPF2e;
+    formDataCache: FormOutputData;
+
+    constructor(
+        object: ActorPF2e,
+        options: Partial<FormApplicationOptions>,
+        callback: CastingItemCreateCallback,
+        spell: SpellPF2e
+    ) {
+        super(object, options);
+
+        this.spell = spell;
+        this.formDataCache = {
+            itemType: this.spell.isCantrip ? "cantripDeck5" : "scroll",
+            level: spell.baseLevel,
+        };
+        this.onSubmitCallback = callback;
+    }
+
+    static override get defaultOptions() {
+        const options = super.defaultOptions;
+
+        options.classes = [];
+        options.title = game.i18n.localize("PF2E.CastingItemCreateDialog.title");
+        options.template = "systems/pf2e/templates/popups/casting-item-create-dialog.hbs";
+        options.width = "auto";
+        options.submitOnChange = true;
+        options.closeOnSubmit = false;
+
+        return options;
+    }
+
+    override async getData(): Promise<FormApplicationData<ActorPF2e>> {
```

Comment:
You should make the FormInputData the return type, and have it extend from FormApplicationData<ActorPF2e>. Rather than setting properties on the super, you should return a new object with the old properties splatted at the start. Most sheet classes more or less follow that pattern now.